### PR TITLE
Contrail R4.1

### PR DESCRIPTION
**Pull request by eonpatapon**

### CODE REVIEW

File: modules/all-in-one.nix [hunk 1]
```diff
@@ -46,7 +67,17 @@ in {
...
+
+      users.users.root = {
+        password = "";
+      } // optionalAttrs cfg.debug {
+        packages = with pkgs; [ gdb ];
+      };
+
+      boot.postBootCommands = mkIf cfg.debug (mkAfter ''
+        cat ${gdbConfig} > /root/.gdbinit
+        ln -s ${contrailPkgs.contrailController} /root/controller
```
Comment (nlewo) [inline modules/all-in-one.nix]: Why this link?

File: modules/configuration/R3.2/vrouter-agent.nix [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    disable_flow_collection = 1
+
+    [DISCOVERY]
+    port = 5998
+    server = ${cfg.discoveryHost}
+
+    [VIRTUAL-HOST-INTERFACE]
+    name = vhost0
+    ip = ${cfg.vhostIP}/24
+    gateway = ${cfg.vhostGateway}
```
Comment (nlewo) [inline modules/configuration/R3.2/vrouter-agent.nix]: Are you sure? Really sure? :/

File: modules/contrail-vrouter-agent.nix [hunk 3]
```diff
@@ -231,14 +214,16 @@ in {
...
+            path = with pkgs; [ netcat iptables ];
             script = ''
               while ! nc -vz localhost 9091; do
                 sleep 2
               done
               ${contrailPkgs.configUtils}/bin/provision_vgw_interface.py --oper create \
                   --interface vgw --subnets ${gw.networkCIDR} --routes ${gw.routes} \
                   --vrf "default-domain:${gw.projectName}:${gw.networkName}:${gw.networkName}"
+            '' + optionalString gw.masquerade ''
+              iptables -t nat -A POSTROUTING -s ${gw.networkCIDR} -j MASQUERADE
```
Comment (nlewo) [inline modules/contrail-vrouter-agent.nix]: Waht about using the firewall NixOS module instead?

File: modules/all-in-one.nix (hunk 1 above)
Comment (eonpatapon) [reply modules/all-in-one.nix]: It's just for easy access to the sources in gdb

File: modules/configuration/R3.2/vrouter-agent.nix (hunk 2 above)
Comment (eonpatapon) [reply modules/configuration/R3.2/vrouter-agent.nix]: :D

File: modules/contrail-vrouter-agent.nix (hunk 3 above)
Comment (eonpatapon) [reply modules/contrail-vrouter-agent.nix]: `networking.firewall.extraCommands` ?

Comment (nlewo) [reply modules/contrail-vrouter-agent.nix]: Yep

File: modules/all-in-one.nix (hunk 1 above)
Comment (nlewo) [reply modules/all-in-one.nix]: It would then be better to add in the gdb config:
```
directory ${contrailPkgs.workspace}
```

**Pull request merged**